### PR TITLE
HHS data:  population-proportionate variants

### PR DESCRIPTION
**Pull request by zhuoran-Cheng16**
### Description
create 2 new signal CONFIRMED_PROP, SUM_CONF_SUSP_PROP that will calculate the proportion of the population of HHS data, with the formula:
CONFIRMED_PROP: **confirmed admissions per day / sum population in state * 100000**
SUM_CONF_SUSP_PROP: **sum of confirmed suspected admissions per day / sum population in state * 100000**

### Changelog
 - Add population column by use state code, create a data frame with new value of the population proportionate value 
 - Add POP_PROP function test and make_signal function test

### CODE REVIEW

File: hhs_hosp/delphi_hhs/run.py [hunk 1]
```diff
@@ -98,7 +98,7 @@ def run_module(params):
     all_columns = pd.concat(dfs)
     geo_mapper = GeoMapper()
     stats = []
-
+    all_columns=geo_mapper.add_population_column(all_columns, "state_code",geocode_col="state")
```
Comment (krivard) [inline hhs_hosp/delphi_hhs/run.py:101]: step through this in a debugger -- before this line, all_columns has a pile of data in it; after this line, it is empty

File: hhs_hosp/delphi_hhs/run.py [hunk 2]
```diff
@@ -139,6 +140,13 @@ def smooth_values(df, smoother):
     )
     return df
 
+def pop_proportion(df):
+    """Get the population-proportionate variants as the dataframe val."""
+    geo_mapper = GeoMapper()
```
Comment (krivard) [inline hhs_hosp/delphi_hhs/run.py:145]: Pass the GeoMapper object in as an argument. GeoMapper has to read the crosswalk files from disk once for each instance, so ideally we only ever make one GeoMapper object.

File: hhs_hosp/delphi_hhs/constants.py [hunk 3]
```diff
@@ -3,11 +3,12 @@
 
 CONFIRMED = "confirmed_admissions_covid_1d"
 SUM_CONF_SUSP = "sum_confirmed_suspected_admissions_covid_1d"
-
+POP_PROP="population_proportionate_variants"
```
Comment (krivard) [inline hhs_hosp/delphi_hhs/constants.py:6]: The basic structure and math looks good!

For signal names we want 
* a variant of each signal, and
* make it clear which signal is the base

so something like this:

```suggestion
CONFIRMED_PROP = "confirmed_admissions_covid_1d_prop"
SUM_CONF_SUSP_PROP = "sum_confirmed_suspected_admissions_covid_1d_prop"
```

& then carry that through to `make_signal`, etc

File: hhs_hosp/tests/test_run.py [hunk 4]
```diff
@@ -54,6 +54,7 @@ def test_make_signal():
     data = pd.DataFrame({
         'state': ['NA'],
         'date': [20200102],
+        'population':[100.],
```
Comment (krivard) [inline hhs_hosp/tests/test_run.py:57]: looks like this column is never used -- be sure to test expected `make_signal` output for all signals.

File: hhs_hosp/delphi_hhs/run.py [hunk 5]
```diff
@@ -176,6 +183,16 @@ def make_signal(all_columns, sig):
...
+    elif sig ==  POP_PROP:
+        df = pd.DataFrame({
+            "state": all_columns.state.apply(str.lower),
+            "timestamp":int_date_to_previous_day_datetime(all_columns.date),
+            "val": \
+            all_columns.previous_day_admission_adult_covid_confirmed + \
+            all_columns.previous_day_admission_adult_covid_suspected + \
+            all_columns.previous_day_admission_pediatric_covid_confirmed + \
+            all_columns.previous_day_admission_pediatric_covid_suspected,
+        })
```
Comment (krivard) [inline hhs_hosp/delphi_hhs/run.py:195]: once the signal names are more normalized you'll be able to drop this block, and instead modify the conditions on the previous if's to something like

```
    if sig.startswith(CONFIRMED):
        ...
    elif sig.startswith(SUM_CONF_SUSP):
        ...
```

**Review by krivard — CHANGES_REQUESTED** [review]
good progress, next fix is signal naming/configuration & test coverage

File: hhs_hosp/delphi_hhs/constants.py (hunk 3 above)
Comment (zhuoran-Cheng16) [reply hhs_hosp/delphi_hhs/constants.py:6]: Got it!

File: hhs_hosp/delphi_hhs/run.py [hunk 6]
```diff
@@ -103,7 +102,7 @@ def run_module(params):
                                     "state_id",
                                     "state_code",
                                     from_col="state")
-        if sensor==POP_PROP:
+        if sensor.endswith("_PROP"):
```
Comment (krivard) [inline hhs_hosp/delphi_hhs/run.py:105]:
```suggestion
        if sensor.endswith("_prop"):
```

File: hhs_hosp/delphi_hhs/run.py [hunk 7]
```diff
@@ -143,7 +142,7 @@ def smooth_values(df, smoother):
 def pop_proportion(df,geo_mapper):
     """Get the population-proportionate variants as the dataframe val."""
     pop_val=geo_mapper.add_population_column(df, "state_code")
-    df["val"]=df["val"]/pop_val["population"]*100000
+    df["val"]=round(df["val"]/pop_val["population"]*100000,2)
```
Comment (krivard) [inline hhs_hosp/delphi_hhs/run.py:145]: we use 7 decimal places systemwide
```suggestion
    df["val"]=round(df["val"]/pop_val["population"]*100000,7)
```

File: hhs_hosp/tests/test_run.py [hunk 8]
```diff
@@ -68,13 +67,27 @@ def test_make_signal():
     })
     pd.testing.assert_frame_equal(expected_confirmed, make_signal(data, CONFIRMED))
 
+    expected_confirmed = pd.DataFrame({
+        'state': ['na'],
+        'timestamp': [datetime(year=2020, month=1, day=1)],
+        'val': [5.],
+    })
```
Comment (krivard) [inline hhs_hosp/tests/test_run.py:74]: identical; no need to set it again
```suggestion
```

File: hhs_hosp/tests/test_run.py [hunk 9]
```diff
@@ -68,13 +67,27 @@ def test_make_signal():
...
         'timestamp': [datetime(year=2020, month=1, day=1)],
         'val': [15.],
     })
     pd.testing.assert_frame_equal(expected_sum, make_signal(data, SUM_CONF_SUSP))
 
+    expected_sum = pd.DataFrame({
+        'state': ['na'],
+        'timestamp': [datetime(year=2020, month=1, day=1)],
+        'val': [15.],
+    })
```
Comment (krivard) [inline hhs_hosp/tests/test_run.py:88]: same
```suggestion
```

**Review by krivard — CHANGES_REQUESTED** [review]
minor substantive changes

File: hhs_hosp/delphi_hhs/run.py [hunk 10]
```diff
@@ -98,12 +97,13 @@ def run_module(params):
...
     geo_mapper = GeoMapper()
     stats = []
-
     for sensor, smoother, geo in product(SIGNALS, SMOOTHERS, GEOS):
         df = geo_mapper.add_geocode(make_signal(all_columns, sensor),
                                     "state_id",
                                     "state_code",
                                     from_col="state")
+        if sensor.endswith("_prop"):
+            df=pop_proportion(df,geo_mapper)
```
Comment (krivard) [inline hhs_hosp/delphi_hhs/run.py:106]:
```suggestion
            df=pop_proportion(df, geo_mapper)
```

File: hhs_hosp/delphi_hhs/run.py [hunk 11]
```diff
@@ -139,6 +139,12 @@ def smooth_values(df, smoother):
     )
     return df
 
+def pop_proportion(df,geo_mapper):
+    """Get the population-proportionate variants as the dataframe val."""
+    pop_val=geo_mapper.add_population_column(df, "state_code")
+    df["val"]=round(df["val"]/pop_val["population"]*100000,7)
```
Comment (krivard) [inline hhs_hosp/delphi_hhs/run.py:145]:
```suggestion
    df["val"]=round(df["val"]/pop_val["population"]*100000, 7)
```

File: hhs_hosp/tests/test_run.py [hunk 12]
```diff
@@ -5,8 +5,8 @@
 import os
 
 from delphi_hhs.run import _date_to_int, int_date_to_previous_day_datetime, generate_date_ranges, \
-    make_signal, make_geo, run_module
-from delphi_hhs.constants import CONFIRMED, SUM_CONF_SUSP, SMOOTHERS, GEOS, SIGNALS
+    make_signal, make_geo, run_module, pop_proportion
+from delphi_hhs.constants import CONFIRMED, SUM_CONF_SUSP, SMOOTHERS, GEOS, SIGNALS,CONFIRMED_PROP, SUM_CONF_SUSP_PROP
```
Comment (krivard) [inline hhs_hosp/tests/test_run.py:9]:
```suggestion
from delphi_hhs.constants import CONFIRMED, SUM_CONF_SUSP, SMOOTHERS, GEOS, SIGNALS, CONFIRMED_PROP, SUM_CONF_SUSP_PROP
```

File: hhs_hosp/tests/test_run.py [hunk 13]
```diff
@@ -66,17 +66,49 @@ def test_make_signal():
...
+        'state': ['WV'],
+        'state_code':[54],
+        'timestamp': [datetime(year=2020, month=1, day=1)],
+        'val': [150.],})
+
+    pd.testing.assert_frame_equal(
+        pop_proportion(test_df,geo_mapper),
+        pd.DataFrame({
+            'state': ['WV'],
+            'state_code':[54],
```
Comment (krivard) [inline hhs_hosp/tests/test_run.py:108]:
```suggestion
        'state_code': [42],
        'timestamp': [datetime(year=2020, month=1, day=1)],
        'val': [15.],})
    pd.testing.assert_frame_equal(
        pop_proportion(test_df, geo_mapper),
        pd.DataFrame({
            'state': ['PA'],
            'state_code': [42],
            'timestamp': [datetime(year=2020, month=1, day=1)],
            'val': [0.1171693],})
    )

    test_df= pd.DataFrame({  
        'state': ['WV'],
        'state_code': [54],
        'timestamp': [datetime(year=2020, month=1, day=1)],
        'val': [150.],})

    pd.testing.assert_frame_equal(
        pop_proportion(test_df, geo_mapper),
        pd.DataFrame({
            'state': ['WV'],
            'state_code': [54],
```

**Review by krivard — APPROVED** [review]
Looks great! These are just whitespace fixes; I'll apply them before I merge

**Pull request merged**